### DISCUSSION TITLE
How to update a texture continuously with a changing Image?

### DISCUSSION THREAD

**Discussion by DoctorWhoof** [emilk/egui#2994] [Q&A]
I have a small library that renders (software only) a texture on every frame.
I was able to get it displaying in an App using macroquad + egui, but I'm trying to get it working with egui only.

I there a way to constantly update a texture with a color buffer, like when using SDL's "streaming textures" or Macroquad's texture.update(image)?

This code snippet below works... but concerns me. Not only I have to clone the image on every frame (load_texture won't let me just borrow the image), the docs explicitly warn against calling load_texture on every frame.

```rust
// self.image is of ColorImage type, and the code that updates it has
// already executed at this point. I just need to display the result.

egui::CentralPanel::default().show(ctx, |ui| {

    // Cloning the image? Yikes, is that going to scale well?
    // Also, load_texture on every frame? Isn't that...bad?
    let texture = ctx.load_texture("test", self.image.clone(), TextureOptions::default());
    let texture_id = TextureId::from(&texture);

    // Draw
    let rect = Rect{
        min: pos2(0.0, 0.0),
        max: pos2(self.res.x, self.res.y)
    };
    let uv = Rect{ min:pos2(0.0, 0.0), max:pos2(1.0, 1.0)};
    ui.painter().image(texture_id, rect, uv, Color32::WHITE);

});
```

Is there a better way?
Or is this Ok? (memory use seems stable, and CPU use is Ok-ish)
Thanks!

Edit: When I split up the image into separate slices (that can be software-rendered in parallel), CPU use is suddenly not Ok anymore, way higher than if I do that part in macroquad alone, maybe due to all the image cloning and texture loading happening per slice, per frame?

**Comment by RasmusHogslatt** [comment]
I have a similar if not the same issue. I made a version of game-of-life using eframe, but it only updates whenever I do some interaction, such as moving the cursor. If I open another application while running, my gui is not doing anything. Is there a way to continuosly keep updating the gui regardless of user interactions? It's as if it's running a retained mode GUI, but I am not able to find any options to switch to continous. I noticed the web demo seemingly has that option. I am using eframe 0.22.0

**Reply by stonerfish** [reply]
The demo uses a variable called "run_mode" to control if a redraw is done each frame.
You do not need to use that variable.

What you do need to do is in your update() function add a call to request_repaint()

```
fn update(&mut self, ctx: &egui::Context, frame: &mut eframe::Frame) {
      ...your code....
     ctx.request_repaint();
}
```
https://docs.rs/egui/0.22.0/egui/struct.Context.html#method.request_repaint

**Reply by RasmusHogslatt** [reply]
Ah, thanks!

**Reply by DoctorWhoof** [reply]
In my case the issue was about updating the image pixels efficiently, on every frame. I can access the pixels and modify them just fine, but then to display using egui I need to clone the image and use load_texture on every frame as well, which the docs warn against, but I can't find any other way.

I also looked into RetainedImage from egui_extras, but the examples don't actually manipulate the image per pixel.

**Reply by WannaBeKratos** [reply]
I suggest you look into displaying the image via wgpu using the native textures, or there is a set method under the [TextureHandle](https://docs.rs/egui/latest/egui/struct.TextureHandle.html). It may be more efficient :
```
self.renderstate.queue.write_texture(
            ImageCopyTexture {
                texture: &self.texture,
                mip_level: 0,
                origin: Origin3d::ZERO,
                aspect: TextureAspect::All,
            },
            self.image.as_raw().as_slice(),
            ImageDataLayout {
                offset: 0,
                bytes_per_row: Some(
                    NonZeroU32::new(
                        (self.image.width() as usize * std::mem::size_of::<image::Rgba<u8>>())
                            as u32,
                    )
                    .unwrap(),
                ), // Assuming 4 channels (RGBA) and each pixel is 32 bits
                rows_per_image: Some(NonZeroU32::new(self.image.height()).unwrap()),
            },
            self.texture.size(),
        );
```

**Reply by stonerfish** [reply]
"there is a set method under the [TextureHandle](https://docs.rs/egui/latest/egui/struct.TextureHandle.html). It may be more efficient :"

Thank you for that.  I have read the docs recursively, but totally missed that method.

**Reply by mynewmail33** [reply]
Thanks! Sounds just right.

**Reply by DoctorWhoof** [reply]
Thanks, and sorry for the delay!

Using "texture_handle.set" does indeed work and prevents using the "load_texture" method from the context on every frame, but it still requires cloning the image that I'm modifying, since .set consumes the image passed to it. I'm wondering if it's possible to avoid that clone and pass a reference instead.

**Reply by tedsteen** [reply]
I have the same question. Would be nice to be able to access the underlying `ColorImage` or atleast the bytes to do the update in place.
